### PR TITLE
perf(recipe): optimize Moonlight training throughput

### PR DESCRIPTION
**Pull request by yaoyu-33**
# What does this PR do ?

Optimize the Moonlight-16B H100 pretrain, SFT, and PEFT recipes using DeepSeek-V3-derived performance settings validated on EOS, refresh the model verification card from real RP2 and Tulu3 workloads, and make equal-token 8K offline packing the default target for future model-card SFT/PEFT verification.

# Changelog

- Increase the 16-GPU pretrain micro-batch size to 2, disable activation recompute, enable router fusion, and use BF16 gradients and optimizer moments while retaining FP32 main parameters.
- Keep each pretrain expert-parallel group within one 8-GPU NVLink domain (EP8 with two expert-data replicas), select HybridEP with the unified 32-SM setting, and overlap MoE communication with delayed weight-gradient compute.
- Configure the TP1 full-SFT recipe without activation recompute and with router fusion plus precision-aware BF16 optimizer state, select HybridEP with overlapped expert communication and delayed weight-gradient compute, and increase offline packs from 2K to 8K while reducing GBS from 32 to 8 to preserve 65,536 token slots per update.
- Move the 4-GPU LoRA/DoRA recipe from TP4/SP-on to TP1/SP-off, disable recompute, and enable router fusion.
- Extend the Moonlight recipe contract tests, update the recipe recommender topology index and Moonlight topology documentation, and refresh the real-data model verification evidence.
- Update the model-card and sequence-packing skills to target 8K/GBS8 for supported offline-packed SFT and PEFT, derive `pad_seq_to_mult` from CP/TP/SP for both workloads, make fixed-width padding conditional on dispatcher/kernel or CUDA-graph requirements, require a DP divisor of eight and fresh convergence evidence, and retain historical verified 2K cards until each model is rerun.

# EOS validation

All pretrain and SFT training measurements below use the public `scripts/training/train.sh` launcher. The final SFT run uses one 8x H100 node; the final pretrain run uses two such nodes.

Pretrain mock-data sweep: 16x H100 80 GB, BF16, sequence length 4096, GBS/MBS 1024/2 for the final recipe, 8 tasks per node, 28 CPUs per task, and exclusive two-node allocations. Reported means exclude the first initialization step.

| Pretrain configuration | Routing | Steady steps | TFLOP/s/GPU |
| --- | --- | ---: | ---: |
| all-to-all, EP16, MBS1 | natural | 9 | 195.87 |
| all-to-all, EP16, MBS2 | force-balanced benchmark | 9 | 258.56 |
| HybridEP, EP16, MBS2 | force-balanced benchmark | 9 | 329.43 |
| HybridEP, EP8, MBS2 | force-balanced benchmark | 9 | 358.11 |
| HybridEP, EP8, MBS2, MoE overlap | natural (recipe default) | 19 | 392.94 (371.4-398.0) |
| HybridEP, EP8, MBS2, MoE overlap | force-balanced benchmark | 19 | 397.87 (395.6-399.1) |

MBS4 was rejected: both no-recompute and selective-recompute variants ran out of 80 GB H100 memory while attempting an additional 5 GiB allocation. The final recipe retains MBS2 and 32-way gradient accumulation. Force-balanced routing was used only for controlled benchmarks and remains disabled in the convergence recipe.

The SFT dispatcher comparison held the real Tulu3 data, TP1/EP8/DP8 topology, MBS1, GBS32, sequence length 2048, and four-way gradient accumulation constant:

| SFT configuration | Steady steps | Step time | TFLOP/s/GPU |
| --- | ---: | ---: | ---: |
| all-to-all with MoE overlap | 10 | 2,080.32 ms | 74.28 |
| HybridEP with MoE overlap and delayed weight-gradient compute | 10 | 1,891.56 ms | 84.78 |

The public launcher also exposed a shared offline-packing output collision when two candidates prepared the same dataset concurrently; the isolated rerun used a unique output root and completed with finite loss.

The equal-token offline-packing sweep then held the real Tulu3 data, TP1/EP8/DP8 topology, MBS1, and 65,536 token slots per optimizer update constant:

| Packed length | Packing efficiency | Source sequences / pack | Gradient accumulation | Last-10 step time | Last-10 TFLOP/s/GPU |
| ---: | ---: | ---: | ---: | ---: | ---: |
| 2,048 | 99.20% | 5.510 | 4 | 1,287.11 ms | 100.00 |
| 4,096 | 99.65% | 10.817 | 2 | 853.42 ms | 151.17 |
| 8,192 | 99.73% | 21.550 | 1 | 562.14 ms | 229.50 |

The final 8K configuration therefore removes gradient accumulation and improves equal-token throughput by 2.29x over the 2K baseline. Fixed-width padding only fills the small residual gap at the end of each offline pack; it is required because the HybridEP kernel processes 128-token chunks.

Real-data verification was rerun from the immutable Bridge commit recorded in the card:

| Mode | Dataset / topology | Result |
| --- | --- | --- |
| Pretrain | RP2, 16x H100, TP1/EP8/DP16, GBS/MBS 1024/2 | 386.590 TFLOP/s/GPU, 11,636.350 ms/step (last 10); loss 12.432480 to 5.158420; 100/100 finite steps. |
| Checkpoint resume | RP2 step 50 to 100, same launch topology | 383.920 TFLOP/s/GPU, 11,717.960 ms/step (last 10); loss 5.775853 to 5.150805; step-100 reference delta 0.007615 (0.148%). |
| SFT | Tulu3, 8x H100, TP1/EP8/DP8, 8K packs, GBS/MBS 8/1 | 229.500 TFLOP/s/GPU, 562.140 ms/step (last 10); loss 1.770799 to 1.180197; 100/100 finite steps. |
| LoRA | Tulu3, 4x H100, TP1/EP4/DP4 | 67.380 TFLOP/s/GPU, 3,822.840 ms/step (last 10); loss 1.264324 to 1.100755; 100/100 finite steps. |
| SFT export / inference | 1x H100 | 27 BF16 shards and 5,344 weights reload strictly; two 32-token greedy generations are byte-identical. |

# Validation

- `uv run --isolated --no-project --with pre-commit==4.6.1 pre-commit run --all-files` - all hooks passed.
- `uv run --active --no-sync python -m pytest tests/unit_tests/recipes/test_moonlight_recipes.py -k sft_convergence_contract -q` - passed in the EOS Linux container (Slurm job 5689097).
- Both updated skill directories pass the skill-creator `quick_validate.py` check, and the sequence-packing evaluation JSON and card YAML parse successfully.
- `uv run --isolated --no-project --with pyyaml python skills/create-model-verification-card/scripts/validate_card.py <card>` - all four checked-in model verification cards are valid.
- The local macOS environment cannot install the Linux-only `nvidia-resiliency-ext` wheel, so final validation used the project container on EOS.

# Before your PR is "Ready for review"

**Pre checks**:

- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA-NeMo/Megatron-Bridge/blob/main/CONTRIBUTING.md)
- [x] Did you write any new necessary tests?
- [x] Did you add or update any necessary documentation?
- [x] Does the PR affect components that are optional to install? No.
  - [x] Reviewer: Does the PR have correct import guards for all optional libraries? Not applicable.

# Additional Information

- Related to MB-897.

### CODE REVIEW

**Comment by yaoyu-33** [comment]
/ok to test https://github.com/NVIDIA-NeMo/Megatron-Bridge/pull/5021/commits/75f04be1fcc1cfab52580315d92b82cb0bb83bfe
